File: backend/src/services/delivery.service.ts
import { DeliveryRepository } from '../repositories/delivery.repository';
import { OrderRepository } from '../repositories/order.repository';
import { emitOrderStatusChange } from '../websocket/socket';

export class DeliveryService {
  private deliveryRepository: DeliveryRepository;
  private orderRepository: OrderRepository;

  constructor() {
    this.deliveryRepository = new DeliveryRepository();
    this.orderRepository = new OrderRepository();
  }

  async acceptDelivery(orderId: string, deliveryPersonId: string) {
    const order = await this.orderRepository.findById(orderId);
    if (!order) throw new Error('Pedido não encontrado.');

    if (order.status === 'RECEIVED') {
      throw new Error('O restaurante ainda não começou a preparar este pedido.');
    }

    const existingDelivery = await this.deliveryRepository.findByOrderId(orderId);
    if (existingDelivery) throw new Error('Esta entrega já foi aceita por outro entregador.');

    // 1. Cria a corrida vinculando ao entregador
    const delivery = await this.deliveryRepository.create({
      orderId,
      deliveryPersonId,
      acceptedAt: new Date(),
    });

    return delivery;
  }

  async pickUpDelivery(orderId: string, deliveryPersonId: string) {
    const delivery = await this.deliveryRepository.findByOrderId(orderId);
    if (!delivery) throw new Error('Entrega não encontrada.');
    if (delivery.deliveryPersonId !== deliveryPersonId) throw new Error('Acesso negado: Esta entrega não é sua.');
    if (delivery.pickedUpAt) throw new Error('Esta entrega já foi retirada.');

    const order = await this.orderRepository.findById(orderId);
    if (order?.status !== 'READY_FOR_PICKUP') {
      throw new Error('Calma! O restaurante ainda não marcou este pedido como "Pronto para Retirada".');
    }

    await this.deliveryRepository.markAsPickedUp(delivery.id);
    await this.orderRepository.updateStatus(orderId, 'ON_THE_WAY');
    emitOrderStatusChange(orderId, 'ON_THE_WAY'); // Só avisa o cliente agora!

    return delivery;
  }

  async markAsDelivered(orderId: string, deliveryPersonId: string, verificationCode: string) {
    const delivery = await this.deliveryRepository.findByOrderId(orderId);
    if (!delivery) throw new Error('Entrega não encontrada.');
    if (delivery.deliveryPersonId !== deliveryPersonId) throw new Error('Acesso negado: Esta entrega não é sua.');
    if (delivery.deliveredAt) throw new Error('Esta entrega já foi finalizada.');
    if (!delivery.pickedUpAt) throw new Error('Você precisa retirar o pedido no restaurante primeiro.');

    const order = await this.orderRepository.findById(orderId);
    if (order?.verificationCode !== verificationCode) {
      throw new Error('Código de verificação incorreto.');
    }

    await this.deliveryRepository.markAsDelivered(delivery.id);
    await this.orderRepository.updateStatus(orderId, 'DELIVERED');
    emitOrderStatusChange(orderId, 'DELIVERED');

    return delivery;
  }

  async listMyDeliveries(deliveryPersonId: string) {
    return this.deliveryRepository.findActiveDeliveriesByPerson(deliveryPersonId);
  }

  async getMyDeliveryHistory(deliveryPersonId: string) {
    return this.deliveryRepository.findDeliveryHistoryByPerson(deliveryPersonId);
  }

  async getAvailableDeliveries() {
    return this.orderRepository.findAvailableForDelivery();
  }
}

File: backend-java/src/main/java/com/foodie/delivery/service/DeliveryService.java
package com.foodie.delivery.service;

import com.foodie.delivery.domain.entity.Delivery;
import com.foodie.delivery.domain.entity.Order;
import com.foodie.delivery.domain.entity.User;
import com.foodie.delivery.domain.enums.DeliveryMethod;
import com.foodie.delivery.domain.enums.OrderStatus;
import com.foodie.delivery.repository.DeliveryRepository;
import com.foodie.delivery.repository.OrderRepository;
import com.foodie.delivery.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class DeliveryService {

    private final DeliveryRepository deliveryRepository;
    private final OrderRepository orderRepository;
    private final UserRepository userRepository;
    private final WebSocketNotificationService webSocketNotificationService;

    @Transactional
    public Delivery acceptDelivery(UUID orderId, UUID deliveryPersonId) {
        Order order = orderRepository.findById(orderId).orElseThrow(() -> new IllegalArgumentException("Pedido não encontrado."));
        
        if (order.getStatus() == OrderStatus.RECEIVED) throw new IllegalArgumentException("O restaurante ainda não começou a preparar este pedido.");
        if (deliveryRepository.findByOrderId(orderId).isPresent()) throw new IllegalArgumentException("Esta entrega já foi aceita por outro entregador.");

        User deliveryPerson = userRepository.findById(deliveryPersonId).orElseThrow(() -> new IllegalArgumentException("Entregador não encontrado."));

        Delivery delivery = Delivery.builder()
                .order(order)
                .deliveryPerson(deliveryPerson)
                .acceptedAt(LocalDateTime.now())
                .build();

        return deliveryRepository.save(delivery);
    }

    @Transactional
    public Delivery pickUpDelivery(UUID orderId, UUID deliveryPersonId) {
        Delivery delivery = deliveryRepository.findByOrderId(orderId).orElseThrow(() -> new IllegalArgumentException("Entrega não encontrada."));

        if (!delivery.getDeliveryPerson().getId().equals(deliveryPersonId)) throw new IllegalArgumentException("Acesso negado: Esta entrega não é sua.");
        if (delivery.getPickedUpAt() != null) throw new IllegalArgumentException("Esta entrega já foi retirada.");

        Order order = delivery.getOrder();
        if (order.getStatus() != OrderStatus.READY_FOR_PICKUP) throw new IllegalArgumentException("Calma! O restaurante ainda não marcou este pedido como \"Pronto para Retirada\".");

        delivery.setPickedUpAt(LocalDateTime.now());
        order.setStatus(OrderStatus.ON_THE_WAY);
        orderRepository.save(order);

        webSocketNotificationService.emitOrderStatusChange(orderId, OrderStatus.ON_THE_WAY);
        return deliveryRepository.save(delivery);
    }

    @Transactional
    public Delivery markAsDelivered(UUID orderId, UUID deliveryPersonId, String verificationCode) {
        Delivery delivery = deliveryRepository.findByOrderId(orderId).orElseThrow(() -> new IllegalArgumentException("Entrega não encontrada."));

        if (!delivery.getDeliveryPerson().getId().equals(deliveryPersonId)) throw new IllegalArgumentException("Acesso negado: Esta entrega não é sua.");
        if (delivery.getDeliveredAt() != null) throw new IllegalArgumentException("Esta entrega já foi finalizada.");
        if (delivery.getPickedUpAt() == null) throw new IllegalArgumentException("Você precisa retirar o pedido no restaurante primeiro.");

        Order order = delivery.getOrder();
        if (!order.getVerificationCode().equals(verificationCode)) throw new IllegalArgumentException("Código de verificação incorreto.");

        delivery.setDeliveredAt(LocalDateTime.now());
        order.setStatus(OrderStatus.DELIVERED);
        orderRepository.save(order);

        webSocketNotificationService.emitOrderStatusChange(orderId, OrderStatus.DELIVERED);
        return deliveryRepository.save(delivery);
    }

    public List<Delivery> listMyDeliveries(UUID deliveryPersonId) { return deliveryRepository.findByDeliveryPersonIdAndDeliveredAtIsNull(deliveryPersonId); }
    public List<Delivery> getMyDeliveryHistory(UUID deliveryPersonId) { return deliveryRepository.findByDeliveryPersonIdAndDeliveredAtIsNotNullOrderByDeliveredAtDesc(deliveryPersonId); }
    
    public List<Order> getAvailableDeliveries() { return orderRepository.findByDeliveryIsNullAndStatusInAndDeliveryMethodOrderByCreatedAtAsc(List.of(OrderStatus.PREPARING, OrderStatus.READY_FOR_PICKUP), DeliveryMethod.DELIVERY); }
}